### PR TITLE
Automatically scatter large arguments in joblib connector

### PR DESCRIPTION
**Pull request by ogrisel**
This requires: https://github.com/joblib/joblib/pull/689 to work efficiently but should not crash or degrade performance when used with an older joblib.

### CODE REVIEW

**Comment by ogrisel** [comment]
@TomAugspurger if you have a chance it would be great to update the docker image of your pangeo instance to test nested parallelism with this new feature (this requires upgrading the joblib branch too).

**Comment by ogrisel** [comment]
I ran some tests on @TomAugspurger's pangeo instance with 30 workers  (randomized search for a random forests) and everything looks fine: the CPUs get saturated and there is very little white space in the task stream view of the diagnostics page of the dask scheduler.

**Comment by ogrisel** [comment]
I also checked that all the futures are garbage collected by the end of the grid search: `client.futures` is empty.

File: distributed/joblib.py [hunk 1]
```diff
@@ -40,6 +41,63 @@
...
 
+def is_weakrefable(obj):
+    try:
+        weakref.ref(obj)
+        return True
+    except TypeError:
+        return False
+
+
+class _WeakKeyDictionary:
```
Comment (jcrist) [inline distributed/joblib.py]: Any reason not to subclass from [MutableMapping](https://docs.python.org/3/library/collections.abc.html#collections.abc.MutableMapping), which would turn this into a full `dict` interface?

File: distributed/joblib.py [hunk 2]
```diff
@@ -116,47 +174,80 @@ def get_nested_backend(self):
...
+                arg_id = id(arg)
+                if arg_id in itemgetters:
+                    yield itemgetters[arg_id]
+                    continue
+
+                f = None
+                if f is None and arg_id in self.data_futures:
+                    f = self.data_futures[arg_id]
+
+                if f is None and call_data_futures is not None:
```
Comment (jcrist) [inline distributed/joblib.py]: Should this be `elif f is None...` instead?

File: distributed/joblib.py [hunk 3]
```diff
@@ -116,47 +174,80 @@ def get_nested_backend(self):
...
+                    try:
+                        f = call_data_futures[arg]
+                    except KeyError:
+                        if (call_data_futures is not None
```
Comment (jcrist) [inline distributed/joblib.py]: `call_data_futures` will always be not `None` due to the `if` check above, this check is not needed.

File: distributed/joblib.py [hunk 4]
```diff
@@ -116,47 +174,80 @@ def get_nested_backend(self):
...
+                                and is_weakrefable(arg)
```
Comment (jcrist) [inline distributed/joblib.py]: IIUC, ` _WeakKeyDictionary` can handle non-weakrefable objects, so this check is not needed?

File: distributed/joblib.py [hunk 5]
```diff
@@ -116,47 +174,80 @@ def get_nested_backend(self):
...
         return self.effective_n_jobs(n_jobs)
 
+    def start_call(self):
+        self.call_data_futures = _WeakKeyDictionary()
+
+    def stop_call(self):
+        # The explicit call to clear is required to break a cycling reference
+        # to the futures.
+        self.call_data_futures.clear()
+        self.call_data_futures = None
```
Comment (jcrist) [inline distributed/joblib.py]: Why set this to `None`? If `call_data_futures` always exists and is just cleared after running, then you never need to check if it's `None` below.

File: distributed/joblib.py (hunk 1 above)
Comment (ogrisel) [reply distributed/joblib.py]: I would need to implement `__delitem__` (easy) and `__iter__` not as simple but we have no use for those. So I don't think it's worth it.

File: distributed/joblib.py (hunk 4 above)
Comment (ogrisel) [reply distributed/joblib.py]: Actually, it's an oversight on my side when implementing the `_WeakKeyDictionary`. I don't think it's possible to safely handle non-weakrefable objects like dict instances: if the id of a recently collected dict argument is reused we will compute the wrong thing. I will make the `_WeakKeyDictionary` refuse to accept non-weakrefable keys.

File: distributed/joblib.py (hunk 5 above)
Comment (ogrisel) [reply distributed/joblib.py]: Alright.

**Comment by ogrisel** [comment]
@jcrist I think I addressed your comments. Let's see of the CI likes it.

**Comment by ogrisel** [comment]
I think the travis failure on `test_administration` is unrelated to this PR.

**Comment by mrocklin** [comment]
Yeah, it looks like a couple of intermittent failures have crept in.
Several PRs are failing today.

**Comment by TomAugspurger** [comment]
CI seems green now.

**Comment by mrocklin** [comment]
I think @jcrist gave a +1 in person.  Merging this now.

**Pull request merged**